File: dotnet/CucumberExpressions/Ast/Token.cs
using System;

namespace CucumberExpressions.Ast;

public class Token : ILocated
{
    public const char EscapeCharacter = '\\';
    public const char AlternationCharacter = '/';
    public const char BeginParameterCharacter = '{';
    public const char EndParameterCharacter = '}';
    public const char BeginOptionalCharacter = '(';
    public const char EndOptionalCharacter = ')';

    public string Text { get; }
    public TokenType Type { get; }
    public int Start { get; }
    public int End { get; }

    public Token(string text, TokenType type, int start, int end)
    {
        Text = text ?? throw new ArgumentNullException(nameof(text));
        Type = type;
        Start = start;
        End = end;
    }

    public static bool CanEscape(char token)
    {
        if (char.IsWhiteSpace(token))
        {
            return true;
        }

        switch (token)
        {
            case EscapeCharacter:
            case AlternationCharacter:
            case BeginParameterCharacter:
            case EndParameterCharacter:
            case BeginOptionalCharacter:
            case EndOptionalCharacter:
                return true;
        }

        return false;
    }

    public static TokenType TypeOf(char token)
    {
        if (char.IsWhiteSpace(token))
        {
            return TokenType.WHITE_SPACE;
        }

        switch (token)
        {
            case AlternationCharacter:
                return TokenType.ALTERNATION;
            case BeginParameterCharacter:
                return TokenType.BEGIN_PARAMETER;
            case EndParameterCharacter:
                return TokenType.END_PARAMETER;
            case BeginOptionalCharacter:
                return TokenType.BEGIN_OPTIONAL;
            case EndOptionalCharacter:
                return TokenType.END_OPTIONAL;
        }

        return TokenType.TEXT;
    }

    public static bool IsEscapeCharacter(int token)
    {
        return token == EscapeCharacter;
    }

    public override string ToString()
    {
        return $"{{\"type\": \"{Type}\", \"start\": \"{Start}\", \"end\": \"{End}\", \"text\": \"{Text}\"}}";
    }

    #region Equality

    protected bool Equals(Token other)
    {
        return Text == other.Text && Type == other.Type && Start == other.Start && End == other.End;
    }

    public override bool Equals(object obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        if (obj.GetType() != this.GetType()) return false;
        return Equals((Token)obj);
    }

    public override int GetHashCode()
    {
        unchecked
        {
            var hashCode = (Text != null ? Text.GetHashCode() : 0);
            hashCode = (hashCode * 397) ^ (int)Type;
            hashCode = (hashCode * 397) ^ Start;
            hashCode = (hashCode * 397) ^ End;
            return hashCode;
        }
    }

    #endregion
}

File: java/src/main/java/io/cucumber/cucumberexpressions/Token.java
package io.cucumber.cucumberexpressions;

import org.jspecify.annotations.Nullable;

import java.util.Objects;
import java.util.StringJoiner;

import static java.util.Objects.requireNonNull;

final class Token implements Located {

    private static final char escapeCharacter = '\\';
    private static final char alternationCharacter = '/';
    private static final char beginParameterCharacter = '{';
    private static final char endParameterCharacter = '}';
    private static final char beginOptionalCharacter = '(';
    private static final char endOptionalCharacter = ')';

    final String text;
    final Type type;
    final int start;
    final int end;

    Token(String text, Type type, int start, int end) {
        this.text = requireNonNull(text);
        this.type = requireNonNull(type);
        this.start = start;
        this.end = end;
    }

    static boolean canEscape(Integer token) {
        if (Character.isWhitespace(token)) {
            return true;
        }
        return switch (token) {
            case (int) escapeCharacter,
                 (int) alternationCharacter,
                 (int) beginParameterCharacter,
                 (int) endParameterCharacter,
                 (int) beginOptionalCharacter,
                 (int) endOptionalCharacter -> true;
            default -> false;
        };
    }

    static Type typeOf(Integer token) {
        if (Character.isWhitespace(token)) {
            return Type.WHITE_SPACE;
        }
        return switch (token) {
            case (int) alternationCharacter -> Type.ALTERNATION;
            case (int) beginParameterCharacter -> Type.BEGIN_PARAMETER;
            case (int) endParameterCharacter -> Type.END_PARAMETER;
            case (int) beginOptionalCharacter -> Type.BEGIN_OPTIONAL;
            case (int) endOptionalCharacter -> Type.END_OPTIONAL;
            default -> Type.TEXT;
        };
    }

    static boolean isEscapeCharacter(int token) {
        return token == escapeCharacter;
    }

    @Override
    public int start() {
        return start;
    }

    @Override
    public int end() {
        return end;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        Token token = (Token) o;
        return start == token.start &&
                end == token.end &&
                text.equals(token.text) &&
                type == token.type;
    }

    @Override
    public int hashCode() {
        return Objects.hash(start, end, text, type);
    }

    @Override
    public String toString() {
        return new StringJoiner(", ", "{", "}")
                .add("\"type\": \"" + type + "\"")
                .add("\"start\": " + start)
                .add("\"end\": " + end)
                .add("\"text\": \"" + text + "\"")
                .toString();
    }

    enum Type {
        START_OF_LINE,
        END_OF_LINE,
        WHITE_SPACE,
        BEGIN_OPTIONAL("" + beginOptionalCharacter, "optional text"),
        END_OPTIONAL("" + endOptionalCharacter, "optional text"),
        BEGIN_PARAMETER("" + beginParameterCharacter, "a parameter"),
        END_PARAMETER("" + endParameterCharacter, "a parameter"),
        ALTERNATION("" + alternationCharacter, "alternation"),
        TEXT;

        private final @Nullable String symbol;
        private final @Nullable String purpose;

        Type() {
            this(null, null);
        }

        Type(@Nullable String symbol, @Nullable String purpose) {
            this.symbol = symbol;
            this.purpose = purpose;
        }

        String purpose() {
            return requireNonNull(purpose, name() + " does not have a purpose");
        }

        String symbol() {
            return requireNonNull(symbol, name() + " does not have a symbol");
        }
    }

}
